### PR TITLE
Better validation messages with SchemaValidator utility

### PR DESCRIPTION
**Pull request by kociolekscramjet**
**What?**
Wild SchemaValidator appeared! SchemaValidator used validation! It's super effective!

For now, the SchemaValidator is used in the ServerConfiguration. Later on, it's going to be used in every *Configuration
class within Orchestrator.And Middleware next.

**Why?**
For more descriptive configuration errors.

**Usage:**
You can build the schema using validators and provide them with proper error messages.

Example schemaValidator:
```ts
const schemaValidator = new SchemaValidator({
    verbose: [optionalValidator, booleanValidator("Server verbose is not valid boolean value")],
    server: [optionalValidator],
    sslKeyPath: [
        optionalValidator,
        stringValidator("Server sslKeyPath is not valid string value"),
        fileExistsValidator("Server sslKeyPath file does not exists or is not readable"),
    ],
    sslCertPath: [
        optionalValidator,
        stringValidator("Server sslCertPath is not valid string value"),
        fileExistsValidator("Server sslCertPath file does not exists or is not readable"),
    ],
    router: [optionalValidator],
});
```
Avaiable validators are here:
```
sth/packages/utility/src/validators/api-version.ts
sth/packages/utility/src/validators/boolean.ts
sth/packages/utility/src/validators/defined.ts
sth/packages/utility/src/validators/fileExists.ts
sth/packages/utility/src/validators/id-string.ts
sth/packages/utility/src/validators/index.ts
sth/packages/utility/src/validators/log-level.ts
sth/packages/utility/src/validators/optional.ts
sth/packages/utility/src/validators/port.ts
sth/packages/utility/src/validators/schema-validator.ts
sth/packages/utility/src/validators/string.ts
sth/packages/utility/src/validators/url-path.ts
```

**How it works:**
The SchemaValidator produces error messages based on the schema. The schemaValidator instance can be used within *Configuration file. It has nicely fitting methods like validate and validateEntry which behave same as those from the ReadOnlyConfig.

Look at the example here https://github.com/scramjetorg/transform-hub/commit/6c0288cfa19bc1ba8192ab1e711ee46443140ce7

**Review checks:**

These aspects need to be checked by the reviewer:

- [x] Verify and confirm operation (please post a screenshot) 
- [x] All STH tests pass
- [x] All [Scramjet Cloud Platform](https://docs.scramjet.org/platform) tests pass
- [x] Documentation is updated or no changes

### CODE REVIEW

File: packages/utility/src/config/readOnlyConfig.ts [hunk 1]
```diff
@@ -22,7 +22,8 @@ export abstract class ReadOnlyConfig<Type extends Object> implements ReadOnlyCon
     }
 
     validate(config: Record<string, any>): boolean {
-        if (Object.keys(config).length === 0) return false;
+        //Commented out - there is no way "any property defined" is proper validation here.
+        //if (Object.keys(config).length === 0) return false;
```
Comment (alicja-gruzdz) [inline packages/utility/src/config/readOnlyConfig.ts:26]: can we remove it?

File: packages/utility/src/validators/schema-validator.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+     */
+    validateSchema(obj: Record<string, any>): ValidationResult[] {
+        this._errors = [];
+
+        for (const key of Object.keys(this.schema)) {
+            const result = this.validateSchemaElement(key, obj[key as keyof Object]);
+
+            if (result === false) continue;
+
+            if (typeof result === "string") {
```
Comment (alicja-gruzdz) [inline packages/utility/src/validators/schema-validator.ts:33]: thanks for adding the info on where validation failed

**Review by alicja-gruzdz — APPROVED** [review]
thanks, this gives us more understanding of validation errors and it's neat the handy validation methods are defined in the utils.

File: packages/utility/src/config/readOnlyConfig.ts (hunk 1 above)
Comment (kociolekscramjet) [reply packages/utility/src/config/readOnlyConfig.ts]: Done :)

File: packages/api-server/src/config/ServerConfiguration.ts [hunk 3]
```diff
@@ -1,32 +1,59 @@
...
+        return schemaValidator;
+    }
+    validate(config: Record<string, any>): boolean {
+        return schemaValidator.validate(config);
+    }
     protected validateEntry(key: string, value: any): boolean | null {
-        return ServerConfiguration.validateEntry(key, value);
+        return schemaValidator.validateEntry(key, value);
     }
     // eslint-disable-next-line complexity
```
Comment (patuwwy) [inline packages/api-server/src/config/ServerConfiguration.ts:55]: not complex anymore

Comment (kociolekscramjet) [reply packages/api-server/src/config/ServerConfiguration.ts]: Done :)

File: packages/utility/src/validators/schema-validator.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                return result;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Validates object
+     * @param obj input object for validation
```
[the remark this answers is not in the document]
Comment (kociolekscramjet) [reply packages/utility/src/validators/schema-validator.ts]: done

[the remark this answers is not in the document]
Comment (kociolekscramjet) [reply packages/utility/src/validators/schema-validator.ts]: and done :)

**Review by patuwwy — APPROVED** [review]

**Pull request merged**